### PR TITLE
schedulers: fix check hot region flow

### LINKED ISSUE #3023: Region heartbeat with unstable Interval will cause statistic problems
1. zero interval will make byteRate and keyRate are NaN.
it will also cause that hot peer is removed early from cache

2. small or large interval will cause statistic is unstable

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

It will fix #3023 and #3198, it also be a part of #3201

It can be reviewd after #3192 and #3202 merged

### What is changed and how it works?

To deal with the problem of unstable heartbeats, this pr introduces three modifications #3023 
1. use the new aot
2. update the hot degree and anticount records after the sum of the intervals has been accumulated for one heartbeat interval, so as to avoid multiple too short heartbeats leading to rapid cooling or rapid heating. At the same time, we avoid that the whole system becomes too sensitive after aot modification.
3. skip the first heartbeat after transfer leader.

To solve the hot degree instability problem caused by inherited peer statistics #3198
1. We record the distribution of peers at the time of heartbeat and compare whether they are already joined in the next heartbeat, and do not put the old && cold peers into the cache anymore.

### Check List

Tests

- Unit test

### CODE REVIEW

**Review by nolouch — COMMENTED** [review]
Could you describe your main changes?

File: pkg/mock/mockcluster/mockcluster.go [hunk 1]
```diff
@@ -317,39 +317,56 @@ func (mc *Cluster) AddLeaderRegionWithReadInfo(
...
 	r = r.Clone(core.SetReadKeys(readKeys))
 	r = r.Clone(core.SetReportInterval(reportInterval))
-	num := mc.HotCache.GetFilledPeriod(statistics.ReadFlow)
-	for i := 0; i < num; i++ {
+	filledNum := 0
+	if len(filledNums) == 0 {
+		filledNum = mc.HotCache.GetFilledPeriod(statistics.WriteFlow)
+	} else {
+		filledNum = filledNums[0]
+	}
```
Comment (disksing) [inline pkg/mock/mockcluster/mockcluster.go:330]:
```
filledNum := mc.HotCache.GetFilledPeriod(statistics.WriteFlow)
if len(filledNums) > 0 {
  filledNum = filledNums[0]
}
```

File: pkg/mock/mockcluster/mockcluster.go [hunk 2]
```diff
@@ -317,39 +317,56 @@ func (mc *Cluster) AddLeaderRegionWithReadInfo(
...
-	num := mc.HotCache.GetFilledPeriod(statistics.WriteFlow)
-	for i := 0; i < num; i++ {
-		items := mc.HotCache.CheckWrite(r)
+
+	filledNum := 0
+	if len(filledNums) == 0 {
+		filledNum = mc.HotCache.GetFilledPeriod(statistics.WriteFlow)
+	} else {
+		filledNum = filledNums[0]
+	}
```
Comment (disksing) [inline pkg/mock/mockcluster/mockcluster.go:359]: ditto.

File: server/statistics/hot_peer_cache_test.go [hunk 3]
```diff
@@ -15,12 +15,18 @@ package statistics
...
+	"time"
 
 	. "github.com/pingcap/check"
 	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/tikv/pd/server/core"
 )
 
+func Test(t *testing.T) {
+	TestingT(t)
+}
```
Comment (disksing) [inline server/statistics/hot_peer_cache_test.go:28]: I've added in another PR.

File: server/statistics/hot_peer.go [hunk 4]
```diff
@@ -26,30 +27,70 @@ const (
 	dimLen
 )
 
+type dimStat struct {
+	typ         int
+	Rolling     *movingaverage.TimeMedian
+	LastAverage *movingaverage.AvgOverTime
```
Comment (disksing) [inline server/statistics/hot_peer.go:33]: what's the difference between the two?

Comment (lhy1024) [reply server/statistics/hot_peer.go:33]: `LastAverage ` is used to control `hot-degree` and `anti-count`, `Rolling` is used to statistic flow.

File: server/statistics/hot_peer.go [hunk 5]
```diff
@@ -41,15 +78,19 @@ type HotPeerStat struct {
...
-	needDelete bool
-	isLeader   bool
-	isNew      bool
+	needDelete         bool
+	isLeader           bool
+	isNew              bool
+	justTransferLeader bool
+	interval           uint64
+	thresholds         [dimLen]float64
+	peers              []uint64
```
Comment (disksing) [inline server/statistics/hot_peer.go:93]: these fields are actually used as parameters, better use another struct to wrap.

Comment (lhy1024) [reply server/statistics/hot_peer.go:93]: yep, but it will change many functions, could we do it in another pr?

**Review by disksing — APPROVED** [review]

File: server/schedulers/utils.go [hunk 6]
```diff
@@ -327,14 +327,16 @@ func (li *storeLoadDetail) toHotPeersStat() *statistics.HotPeersStat {
 	peers := make([]statistics.HotPeerStat, 0, len(li.HotPeers))
 	var totalBytesRate, totalKeysRate float64
 	for _, peer := range li.HotPeers {
-		peers = append(peers, *peer.Clone())
-		totalBytesRate += peer.ByteRate
-		totalKeysRate += peer.KeyRate
+		if peer.HotDegree > 0 {
+			peers = append(peers, *peer.Clone())
+      totalBytesRate += peer.ByteRate
```
Comment (nolouch) [inline server/schedulers/utils.go:332]: ci failed.

**Review by nolouch — COMMENTED** [review]
Cloud you describe the changes how to solve the problems in detail?

File: server/schedulers/hot_region.go [hunk 7]
```diff
@@ -804,13 +804,13 @@ func (bs *balanceSolver) calcProgressiveRank() {
 		greatDecRatio, minorDecRatio := bs.sche.conf.GetGreatDecRatio(), bs.sche.conf.GetMinorGreatDecRatio()
 		switch {
 		case byteHot && byteDecRatio <= greatDecRatio && keyHot && keyDecRatio <= greatDecRatio:
-			// Both byte rate and key rate are balanced, the best choice.
+			// If use the case, both byte rate and key rate will be more balanced, the best choice.
 			rank = -3
 		case byteDecRatio <= minorDecRatio && keyHot && keyDecRatio <= greatDecRatio:
-			// Byte rate is not worsened, key rate is balanced.
+			// If use the case, byte rate will be not worsened, key rate will be more balanced.
```
Comment (rleungx) [inline server/schedulers/hot_region.go:810]:
```suggestion
			// If belong to the case, byte rate will be not worsened, key rate will be more balanced.
```
Or simply no need to change

File: server/schedulers/hot_test.go [hunk 8]
```diff
@@ -1058,3 +1058,75 @@ func addRegionInfo(tc *mockcluster.Cluster, rwTy rwType, regions []testRegionInf
...
+	// some peers are hot, and some are cold #3198
+	rate := uint64(512 * KB)
+	for i := 0; i < statistics.TopNN; i++ {
+		for j := 0; j < statistics.DefaultAotSize; j++ {
+			tc.AddLeaderRegionWithWriteInfo(uint64(i+100), 1, rate*statistics.RegionHeartBeatReportInterval, 0, statistics.RegionHeartBeatReportInterval, []uint64{2, 3}, 1)
+		}
+	}
+	items := tc.AddLeaderRegionWithWriteInfo(201, 1, rate*statistics.RegionHeartBeatReportInterval, 0, statistics.RegionHeartBeatReportInterval, []uint64{2, 3}, 1)
+	c.Check(items[0].GetThresholds()[0], Equals, float64(rate)*statistics.HotThresholdRatio)
+	// Threshold of store 1,2,3 is 409.6 KB and others are 1 KB
```
Comment (rleungx) [inline server/schedulers/hot_test.go:1121]: Can you explain more?

File: server/statistics/hot_peer.go [hunk 9]
```diff
@@ -26,6 +27,42 @@ const (
...
 )
 
+type dimStat struct {
+	typ         int
+	Rolling     *movingaverage.TimeMedian
+	LastAverage *movingaverage.AvgOverTime
+}
+
+func newDimStat(typ int) *dimStat {
+	reportInterval := time.Duration(RegionHeartBeatReportInterval) * time.Second
```
Comment (rleungx) [inline server/statistics/hot_peer.go:37]: directly use `RegionHeartBeatReportInterval * time.Second`?

File: server/statistics/hot_peer.go [hunk 10]
```diff
@@ -26,6 +27,42 @@ const (
 	dimLen
 )
 
+type dimStat struct {
+	typ         int
+	Rolling     *movingaverage.TimeMedian
+	LastAverage *movingaverage.AvgOverTime
```
Comment (rleungx) [inline server/statistics/hot_peer.go:33]: Please add more comments about what they are used for

File: server/statistics/hot_peer.go [hunk 11]
```diff
@@ -110,3 +156,30 @@ func (stat *HotPeerStat) Clone() *HotPeerStat {
...
+	return stat.rollingByteRate.isHot(thresholds) || stat.rollingKeyRate.isHot(thresholds)
+}
+
+func (stat *HotPeerStat) clearLastAverage() {
+	stat.rollingByteRate.clearLastAverage()
+	stat.rollingKeyRate.clearLastAverage()
+}
+
+// Log is used to output some info
+func (stat *HotPeerStat) Log(str string, level func(msg string, fields ...zap.Field)) {
```
Comment (rleungx) [inline server/statistics/hot_peer.go:170]: looks wired

File: server/statistics/hot_peer_cache.go [hunk 12]
```diff
@@ -18,18 +18,23 @@ import (
...
 	"github.com/tikv/pd/server/core"
+	"go.uber.org/zap"
 )
 
 const (
-	topNN             = 60
+	// TopNN is the threshold which means we can get hot threshold from store.
+	TopNN = 60
+	// HotThresholdRatio is used to calculate hot thresholds
+	HotThresholdRatio = 0.8
```
Comment (rleungx) [inline server/statistics/hot_peer_cache.go:31]: no need to make them public

File: server/statistics/hot_peer_cache.go [hunk 13]
```diff
@@ -310,42 +373,75 @@ func (f *hotPeerCache) isRegionHotWithPeer(region *core.RegionInfo, peer *metapb
...
+			}
+			newItem.HotDegree = 1
+			newItem.AntiCount = hotRegionAntiCount
+		}
 		newItem.isNew = true
+		newItem.rollingByteRate = newDimStat(byteDim)
+		newItem.rollingKeyRate = newDimStat(keyDim)
+		newItem.rollingByteRate.Add(bytes, interval)
+		newItem.rollingKeyRate.Add(keys, interval)
+		if newItem.rollingKeyRate.isFull() {
```
Comment (rleungx) [inline server/statistics/hot_peer_cache.go:401]: When will it happen?

File: server/statistics/hot_peer_cache.go (hunk 12 above)
Comment (lhy1024) [reply server/statistics/hot_peer_cache.go:31]: it's used in hot_test.go

Comment (rleungx) [reply server/statistics/hot_peer_cache.go:31]: prefer to use the constant in tests

File: server/statistics/hot_peer.go (hunk 11 above)
Comment (lhy1024) [reply server/statistics/hot_peer.go:170]: Well, I will remove this log temporarily and will add after usb log merged

File: server/statistics/hot_peer_cache.go (hunk 13 above)
Comment (lhy1024) [reply server/statistics/hot_peer_cache.go:401]: hmmm?

File: server/statistics/hot_peer_cache.go [hunk 14]
```diff
@@ -173,12 +190,18 @@ func (f *hotPeerCache) CheckRegionFlow(region *core.RegionInfo) (ret []*HotPeerS
 			}
 		}
 
-		newItem = f.updateHotPeerStat(newItem, oldItem, bytes, keys, time.Duration(interval))
+		newItem = f.updateHotPeerStat(newItem, oldItem, bytes, keys, time.Duration(interval)*time.Second, thresholds)
```
Comment (nolouch) [inline server/statistics/hot_peer_cache.go:193]: the param `thresholds` already contains in `newItem.thresholds`

File: server/statistics/hot_peer_cache.go [hunk 15]
```diff
@@ -310,42 +371,75 @@ func (f *hotPeerCache) isRegionHotWithPeer(region *core.RegionInfo, peer *metapb
...
+	newItem.rollingByteRate.Add(bytes, interval)
+	newItem.rollingKeyRate.Add(keys, interval)
+
+	if !newItem.rollingKeyRate.isFull() {
+		// not update hot degree and anti count
+		newItem.HotDegree = oldItem.HotDegree
+		newItem.AntiCount = oldItem.AntiCount
+	} else {
+		if f.isOldColdPeer(oldItem, newItem.StoreID) {
+			if newItem.isHot(thresholds) {
```
Comment (nolouch) [inline server/statistics/hot_peer_cache.go:424]: item contains thresholds.

**Review by nolouch — COMMENTED** [review]
rest lgtm

**Review by nolouch — APPROVED** [review]
lgtm

File: pkg/movingaverage/time_median.go [hunk 16]
```diff
@@ -29,11 +29,10 @@ type TimeMedian struct {
 }
 
 // NewTimeMedian returns a TimeMedian with given size.
-func NewTimeMedian(aotSize, mfSize, reportInterval int) *TimeMedian {
-	interval := time.Duration(aotSize*reportInterval) * time.Second
+func NewTimeMedian(aotSize, mfSize int, reportInterval time.Duration) *TimeMedian {
 	return &TimeMedian{
-		aotInterval: interval,
-		aot:         NewAvgOverTime(interval),
+		aotInterval: reportInterval,
```
Comment (rleungx) [inline pkg/movingaverage/time_median.go:34]: It seems this value is not the same as before?

**Pull request merged**